src/cluster_import.c

```diff
@@ -0,0 +1,1915 @@
+    rioWriteBulkString(rdb, "SYNCSLOTS", 9);
+    rioWriteBulkString(rdb, "SNAPSHOT-EOF", 12);
+    return retval;
+}
+
+void slotExportLinkBeginSnapshot(slotMigrationLink *link) {
+    connection **conns = zmalloc(sizeof(connection *));
+    *conns = link->client->conn;
+    serverLog(LL_NOTICE,
+              "Beginning snapshot of slot export link %.40s for slots (%s) to target %.40s.",
+              link->linkname,
+              link->slot_ranges_str,
+              link->nodename);
+    if (saveSnapshotToConnectionSockets(conns, 1, 1, 0, childSnapshotForSyncSlot, 1,
+                                        link->slot_ranges) != C_OK) {
+        serverLog(LL_WARNING,
+                  "Slot export link %.40s failed to start slot export for slots (%s)",
+                  link->linkname,
+                  link->slot_ranges_str);
+        finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED, "Failed to start snapshot");
+        return;
+    }
+    updateSlotMigrationLinkState(link, SLOT_EXPORT_SNAPSHOTTING);
+    if (server.debug_pause_after_fork) debugPauseProcess();
+}
+
+void clusterHandleSlotExportBackgroundSaveDone(int bgsaveerr) {
+    listIter li;
+    listNode *ln;
+    listRewind(server.cluster->slot_migration_links, &li);
+    while ((ln = listNext(&li))) {
+        slotMigrationLink *link = (slotMigrationLink *)ln->value;
+        if (link->type != SLOT_MIGRATION_EXPORT) continue;
+        if (link->state != SLOT_EXPORT_SNAPSHOTTING) {
+            continue;
+        }
+        if (bgsaveerr == C_OK) {
+            serverLog(LL_NOTICE,
+                      "Finished snapshotting slots (%s) to target %.40s, beginning incremental stream...",
+                      link->slot_ranges_str,
+                      link->nodename);
+            slotExportBeginStreaming(link);
+        } else {
+            serverLog(LL_WARNING,
+                      "Failed to snapshot slots (%s) to target %.40s",
+                      link->slot_ranges_str,
+                      link->nodename);
+            finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED, "Failed to perform snapshot");
+        }
+    }
+}
+
+int clusterIsSlotExportReadyForReplData(client *c) {
+    if (!c->flag.slot_export_target || !c->slot_migration_link) {
+        return 1;
+    }
+    slotMigrationLink *link = (slotMigrationLink *)c->slot_migration_link;
+    return link->state != SLOT_EXPORT_SNAPSHOTTING &&
+           link->state != SLOT_EXPORT_WAITING_TO_SNAPSHOT;
+}
+
+void clusterFeedSlotExportLinks(int dbid, robj **argv, int argc) {
+    UNUSED(dbid);
+    int i, error_code;
+    int slot = -1;
+    listIter li;
+    listNode *ln;
+
+    /* This function may be called after the command is executed.
+     * At this time, the arg in argv may be rewritten and the encoding
+     * may be an INT. In this case, we need to decode it into a string
+     * object because in getKeysFromCommand all the args are a string. */
```

Comment:
I'm 99% sure we never rewrite it as an integer encoding anymore. We did do it in the past, but I'm fairly sure it's never done anymore. But even so, we have a function `feedReplicationBufferWithObject` which handles this properly in the replication path to avoid the re-allocation (which is slow).

Reply:
Noted. Yeah I think this function could go through an optimization pass. Probably we need to extract some common functionality out of `replication.c`. Obviously feeding replication backlog (which is a shared global backlog) and migration client output buffer will look slightly different but there are some optimizations we should share

Reply:
If we are going to optimize it in a later PR, can we make an issue?